mongo/client_encryption.go

```diff
@@ -150,6 +216,177 @@ func (ce *ClientEncryption) Close(ctx context.Context) error {
+func (ce *ClientEncryption) DeleteKey(ctx context.Context, id primitive.Binary) (*DeleteResult, error) {
+	return executeCEDeleteResult(ce, &id, func(coll *Collection, query []byte) (*DeleteResult, error) {
+		return coll.DeleteOne(ctx, query)
+	})
+}
+
+// GetKeyByAltName returns a key document in the key vault collection with the given keyAltName.
+func (ce *ClientEncryption) GetKeyByAltName(ctx context.Context, keyAltName string) *SingleResult {
+	return executeCESingleResult(ce, nil, func(coll *Collection, _ []byte) *SingleResult {
+		filter := bsoncore.NewDocumentBuilder().AppendString("keyAltNames", keyAltName).Build()
+		return coll.FindOne(ctx, filter)
+	})
+}
+
+// GetKey finds a single key document with the given UUID (BSON binary subtype 0x04). Returns the result of the
+// internal find() operation on the key vault collection.
+func (ce *ClientEncryption) GetKey(ctx context.Context, id primitive.Binary) *SingleResult {
+	return executeCESingleResult(ce, &id, func(coll *Collection, query []byte) *SingleResult {
+		return coll.FindOne(ctx, query)
+	})
+}
+
+// GetKeys finds all documents in the key vault collection. Returns the result of the internal find() operation on the
+// key vault collection.
+func (ce *ClientEncryption) GetKeys(ctx context.Context) (*Cursor, error) {
+	coll, err := ce.majorityReadCollection()
+	if err != nil {
+		return newEmptyCursor(), nil
+	}
+	return coll.Find(ctx, bson.D{})
+}
+
+// RemoveKeyAltName removes a keyAltName from the keyAltNames array of the key document in the key vault collection with
+// the given UUID (BSON binary subtype 0x04). Returns the previous version of the key document.
+func (ce *ClientEncryption) RemoveKeyAltName(ctx context.Context, id primitive.Binary,
+	keyAltName string) *SingleResult {
+
+	return executeCESingleResult(ce, &id, func(coll *Collection, query []byte) *SingleResult {
+		update := bson.A{bson.D{{"$set", bson.D{{"keyAltNames", bson.D{{"$cond", bson.A{bson.D{{"$eq",
+			bson.A{"$keyAltNames", bson.A{keyAltName}}}}, "$$REMOVE", bson.D{{"$filter",
+			bson.D{{"input", "$keyAltNames"}, {"cond", bson.D{{"$ne", bson.A{"$$this", keyAltName}}}}}}}}}}}}}}}
+		return coll.FindOneAndUpdate(ctx, query, update)
+	})
+}
+
+// setRewrapManyDataKeyWriteModels will prepare the WriteModel slice for a bulk updating rewrapped documents.
+func setRewrapManyDataKeyWriteModels(rewrappedDocuments []bsoncore.Document, writeModels *[]WriteModel) error {
+	const idKey = "_id"
+	const keyMaterial = "keyMaterial"
+	const masterKey = "masterKey"
+
+	// Append a slice of WriteModel with the update document per each rewrappedDoc _id filter.
+	for _, rewrappedDocument := range rewrappedDocuments {
+		// Rebuild the document to remove the immutable _id object.
+		rewrappedDocElems, err := rewrappedDocument.Elements()
+		if err != nil {
+			return err
+		}
+		mutableRewrappedDocBuilder := bsoncore.NewDocumentBuilder()
```

Comment:
The _mutable_ prefix is a little strange and may be unnecessary? Can we just say `rewrappedDocBuilder`, `rewrappedDoc`, `docMasterKey`, etc.?

Reply:
It is specifically added to avoid misunderstanding what that version of the rewrapped doc is doing.  Let me digest renaming it, adding a comment so I don't forget.